File: shapes/traditional/src/main/resources/util/ui.js
$(function() {
  window.UI = function(
      figureTransformed, figureSelected, removeFigure, colorChanged,
      addRectangleEvent, addCircleEvent, addTriangleEvent) {
    var self = this
    self.selectedFigure = null
    self.color = ""

    function applyGeneralFigureProperties(figure, obj) {
      obj.set({
        originX: "center",
        originY: "center",
        fill: figure.color,
        left: figure.position.x,
        top: figure.position.y,
        scaleX: figure.transformation.scaleX,
        scaleY: figure.transformation.scaleY,
        angle: figure.transformation.angle})
    }

    function applyFigureProperties(figure, obj) {
      if (figure.shape.$type === "Rect")
        obj.set({ width: figure.shape.width, height: figure.shape.height })
      else if (figure.shape.$type === "Circle")
        obj.set({ radius: figure.shape.radius })
      else if (figure.shape.$type === "Triangle")
        obj.set({ width: figure.shape.width, height: figure.shape.height })

      applyGeneralFigureProperties(figure, obj)
      obj.setCoords()
    }


    function render(figures) {
      var ids = figures.map(function(figure) { return figure.id })
      var map = {}

      ui.canvas.forEachObject(function(obj) {
        if (ids.indexOf(obj.figure.id) != -1 )
          map[obj.figure.id] = obj
        else
          ui.canvas.remove(obj)
      })

      figures.forEach(function(figure) {
        var obj = map[figure.id]
        if (obj) {
          if (!sameFigure(obj.figure, figure)) {
            applyFigureProperties(figure, obj)
            obj.figure = figure
          }
        }
        else {
          if (figure.shape.$type === "Rect")
            obj = new fabric.Rect
          else if (figure.shape.$type === "Circle")
            obj = new fabric.Circle
          else if (figure.shape.$type === "Triangle")
            obj = new fabric.Triangle

          applyFigureProperties(figure, obj)
          obj.figure = figure

          ui.canvas.add(obj)
        }
      })

      ui.canvas.renderAll()
    }

    function objectsModified(options) {
      var obj = options.target
      var figure = obj.figure

      var position = createPosition(obj.left, obj.top)
      var transformation = createTransformation(obj.scaleX, obj.scaleY, obj.angle)

      if (!samePosition(figure.position, position) ||
          !sameTransformation(figure.transformation, transformation))
        figureTransformed({ position: position, transformation: transformation })
    }

    function selectionChanged(options) {
      if (self.selectedFigure == null ||
          !sameFigure(self.selectedFigure, options.target.figure)) {
        self.selectedFigure = options.target.figure
        figureSelected(self.selectedFigure)
      }
    }

    function selectionCleared(options) {
      if (self.selectedFigure != null) {
        self.selectedFigure = null
        figureSelected(self.selectedFigure)
      }
    }


    $(window).keyup(function(event) {
      if (event.keyCode === 46)
        removeFigure()
    })

    ui.colorpicker.on("changeColor", function(event) {
      var color = event.color.toString("hsla")
      if (color != self.color) {
        self.color = color
        colorChanged(self.color)
      }
    })

    self.color = ui.colorpicker.colorpicker("getValue").toString()

    ui.addRectangle.on("click", function() { addRectangleEvent() })
    ui.addCircle.on("click", function() { addCircleEvent() })
    ui.addTriangle.on("click", function() { addTriangleEvent() })

    ui.canvas.renderOnAddRemove = false

    ui.canvas.on({
      "object:modified": objectsModified,
      "object:selected": selectionChanged,
      "selection:cleared": selectionCleared
    })


    self.setColor = function(changeColor) {
      ui.colorpicker.colorpicker("setValue", changeColor)
    }

    self.setFigures = function(figures) {
      if (self.selectedFigure != null)
        for (var i = 0, l = figures.length; i < l; ++i)
          if (figures[i].id === self.selectedFigure.id) {
            self.selectedFigure = figures[i]
            break
          }

      render(figures)
    }
  }
})

File: shapes/src/main/scala/shapes/util/UI.scala
package shapes
package util

import rescala.Var
import rescala.Signal
import rescala.events.Event
import rescala.events.ImperativeEvent
import rescala.events.emptyevent
import makro.SignalMacro.{SignalM => Signal}

import scala.scalajs.js.Array
import scala.scalajs.js.Function1
import scala.scalajs.js.Dynamic
import scala.scalajs.js.Dynamic.global
import scala.scalajs.js.Dynamic.literal
import scala.scalajs.js.Dynamic.newInstance

class UI {
  private var _figures: Signal[List[Figure]] = Signal { List.empty[Figure] }
  private var _changeColor: Event[String] = emptyevent

  private val colorVar = Var("#000000")
  private val selectedFigureVar = Var(Option.empty[Figure])
  private val figureSelectedEvent = new ImperativeEvent[Figure]
  private val figureTransformedEvent = new ImperativeEvent[(Position, Transformation)]
  private val addRectangleEvent = new ImperativeEvent[Unit]
  private val addCircleEvent = new ImperativeEvent[Unit]
  private val addTriangleEvent = new ImperativeEvent[Unit]
  private val removeFigureEvent = new ImperativeEvent[Unit]

  val color: Signal[String] = colorVar
  val selectedFigure: Signal[Option[Figure]] = selectedFigureVar
  val figureSelected: Event[Figure] = figureSelectedEvent
  val figureTransformed: Event[(Position, Transformation)] = figureTransformedEvent
  val addRectangle: Event[Unit] = addRectangleEvent
  val addCircle: Event[Unit] = addCircleEvent
  val addTriangle: Event[Unit] = addTriangleEvent
  val removeFigure: Event[Unit] = removeFigureEvent

  private def applyGeneralFigureProperties(figure: Figure, obj: Dynamic) = {
    obj set literal(
      originX = "center",
      originY = "center",
      fill = figure.color,
      left = figure.position.x,
      top = figure.position.y,
      scaleX = figure.transformation.scaleX,
      scaleY = figure.transformation.scaleY,
      angle = figure.transformation.angle)
  }

  private def applyFigureProperties(figure: Figure, obj: Dynamic) = {
    figure.shape match {
      case Rect(width, height) =>
        obj set literal(width = width, height = height)
      case Circle(radius) =>
        obj set literal(radius = radius)
      case Triangle(width, height) =>
        obj set literal(width = width, height = height)
    }

    applyGeneralFigureProperties(figure, obj)
    obj.setCoords()
  }


  private def render(figures: List[Figure]) = {
    val ids = (figures map { _.id }).toSet[Long]
    val map = collection.mutable.Map.empty[Long, Dynamic]

    global.ui.canvas forEachObject { obj: Dynamic =>
      val figure = obj.figure.asInstanceOf[Figure]
      if (ids contains figure.id)
        map += figure.id -> obj
      else
        global.ui.canvas remove obj
    }

    figures foreach { figure =>
      (map get figure.id) match {
        case Some(obj) =>
          if (obj.figure.asInstanceOf[Figure] != figure) {
            applyFigureProperties(figure, obj)
            obj.figure = figure.asInstanceOf[Dynamic]
          }

        case _ =>
          val obj = figure.shape match {
            case Rect(_, _) => newInstance(global.fabric.Rect)()
            case Circle(_) => newInstance(global.fabric.Circle)()
            case Triangle(_, _) => newInstance(global.fabric.Triangle)()
          }

          applyFigureProperties(figure, obj)
          obj.figure = figure.asInstanceOf[Dynamic]

          global.ui.canvas add obj
      }
    }

    global.ui.canvas.renderAll()
  }

  private def createPosition(obj: Dynamic) =
    Position(
      obj.left.asInstanceOf[Double],
      obj.top.asInstanceOf[Double])

  private def createTransformation(obj: Dynamic) =
    Transformation(
      obj.scaleX.asInstanceOf[Double],
      obj.scaleY.asInstanceOf[Double],
      obj.angle.asInstanceOf[Double])

  private def objectsModified(options: Dynamic) = {
    val obj = options.target
    val figure = obj.figure.asInstanceOf[Figure]

    val position = createPosition(obj)
    val transformation = createTransformation(obj)

    if (figure.position != position || figure.transformation != transformation)
      figureTransformedEvent((position, transformation))
  }

  private def selectionChanged(options: Dynamic) = {
    val figure = options.target.figure.asInstanceOf[Figure]
    selectedFigureVar() = Some(figure)
    figureSelectedEvent(figure)
  }

  private def selectionCleared(options: Dynamic) =
    selectedFigureVar() = None


  global $ { () =>
    (global $ global) keyup { event: Dynamic =>
      if (event.keyCode == 46)
        removeFigureEvent(())
    }


    global.ui.colorpicker on ("changeColor", { event: Dynamic =>
      colorVar() = event.color.toHex().toString()
    })

    colorVar() = (global.ui.colorpicker colorpicker "getValue").toString()


    global.ui.addRectangle on ("click", () => addRectangleEvent(()))
    global.ui.addCircle on ("click", () => addCircleEvent(()))
    global.ui.addTriangle on ("click", () => addTriangleEvent(()))


    global.ui.canvas.renderOnAddRemove = false

    global.ui.canvas on literal(
      "object:modified" -> objectsModified _,
      "object:selected" -> selectionChanged _,
      "selection:cleared" -> selectionCleared _
    )
  }


  def changeColor = _changeColor

  def changeColor_=(changeColor: Event[String]) = global $ { () =>
    _changeColor = changeColor
    _changeColor += { global.ui.colorpicker colorpicker ("setValue", _) }
  }


  def figures = _figures
  
  def figures_=(figures: Signal[List[Figure]]) = global $ { () =>
    _figures = figures

    figures.changed += { figures =>
      selectedFigureVar() = selectedFigureVar.get match {
        case Some(selectedFigure) =>
          figures collectFirst {
            case figure if figure.id == selectedFigure.id => figure
          }
        case _ =>
          None
      }

      render(figures)
    }

    render(figures.get)
  }
}
